### PR TITLE
 Add generic typehint to RepositoryInterface and FactoryInterface

### LINKED ISSUE #453: Add generic typehint to RepositoryInterface and FactoryInterface
Pull request draft: https://github.com/Sylius/SyliusResourceBundle/pull/454

**Describe the proposed solution**

Proposed version for this change: 1.10

I want to have typehinted generic Repositories and Factories in my projects to archive better integration with PhpStan and PhpStorm. Right now when I extend RepositoryInterface with my custom repository I get PhpStan error:

```
  Line   PublicationBundle/src/Repository/PageRepository.php                                    
 ------ --------------------------------------------------------------------------------------- 
  13     Class Doctrine\ORM\EntityRepository specifies template type TEntityClass of interface  
         Doctrine\Persistence\ObjectRepository as                                               
         MyProject\Bundle\PublicationBundle\Entity\PageInterface but it's already specified as   
         object.
```

I would like to add generic typehints to these classes:

* `src/Component/Repository/RepositoryInterface.php`
  add
  ```php
  use Doctrine\Persistence\ObjectRepository;
  
  /**
   * @template T
   * @extends ObjectRepository<T>
   */
  interface RepositoryInterface extends ObjectRepository
  {
  ```
* `src/Component/Factory/FactoryInterface.php`
  add
  ```php
  /**
   * @template T of object
   */
  interface FactoryInterface
  {
      /**
       * @return T
       */
      public function createNew();
  }
  ```
With these changes I can extend generic RepositoryInterface in my custom repository

```php
use Sylius\Component\Resource\Repository\RepositoryInterface;

/** @extends RepositoryInterface<AdminUserInterface> */
interface AdminUserRepositoryInterface extends RepositoryInterface
{
    /** @throws EntityNotFoundException */
    public function findEnabledByEmail(string $email): AdminUserInterface;
}
```

and I can use them such as

```php
class ImageController
{
    /** @param FactoryInterface<ImageInterface> $imageFactory */
    public function __construct(
        private FactoryInterface $imageFactory,
    ) {
    }

    // ...
}
```

or

```php
class ImageStorageService implements ImageStorageServiceInterface
{
    /**
     * @param EntityRepository<ImageVariantTypeInterface> $imageVariantTypeRepository
     * @param FactoryInterface<ImageVariantInterface> $imageVariantFactory
     * @param FactoryInterface<FileInterface> $fileFactory
     */
    public function __construct(
        private EntityRepository $imageVariantTypeRepository,
        private FileStorageServiceInterface $fileStorageService,
        private FactoryInterface $imageVariantFactory,
        private FactoryInterface $fileFactory,
        private ImageVariantServiceInterface $imageVariantService,
    ) {
    }

    // ...
}
```

**Describe alternatives you've considered**

I can avoid these problems with typehinting everything every time when I use Repositories or Factories. I think it would be much better to just typehint repositories in constructor docblock or directlly extend generic types.

I can also use something like https://github.com/cweagans/composer-patches and ~~hack~~ add these types myself without need to fork this repository.

### PR DESCRIPTION
**Pull request by MrSrsen**
| Q               | A
| --------------- | -----
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes https://github.com/Sylius/SyliusResourceBundle/issues/453
| License         | MIT

### CODE REVIEW

File: src/Component/Repository/RepositoryInterface.php [hunk 1]
```diff
@@ -16,6 +16,10 @@
 use Doctrine\Persistence\ObjectRepository;
 use Sylius\Component\Resource\Model\ResourceInterface;
 
+/**
+ * @template T
+ * @extends ObjectRepository<T>
+ */
 interface RepositoryInterface extends ObjectRepository
```
Comment (mbabker) [inline src/Component/Repository/RepositoryInterface.php:23]: I'd add the generic annotations to the methods in this interface too.

`createPaginator()` returns `iterable<T>`, and the parameter for `add()` and `remove()` is `T`

Comment (MrSrsen) [reply src/Component/Repository/RepositoryInterface.php:23]: Yeah, good idea. Thanks.

**Review by lchrusciel — APPROVED** [review]
Would be great to add templates to the implementation of these interfaces as well https://phpstan.org/blog/generics-by-examples#accept-a-class-string-and-return-the-object-of-that-type

**Comment by lchrusciel** [comment]
Hey Martin! Better static analysis is always a plus! Thanks for your proposal and PR! 🎉

**Pull request merged**